### ISSUE TITLE
Localization Depth

### ISSUE THREAD

**Issue by mehalick** [mehalick/Xaki#2]
Need to support localization depth, for example localizing only the target entity (shallow), the target entity and its immediate children (one-level), or a target entity's full object graph (deep).

This feature should mimic an original feature in OrangeJetpack.Localization: [LocalizationDepth.cs](https://github.com/mehalick/OrangeJetpack.Localization/blob/master/src/OrangeJetpack.Localization/LocalizationDepth.cs).

The API would expand to support something similar to:

```csharp
_localizer.Localize<Planet>(planets); // defaults to LocalizationDepth.Shallow
_localizer.Localize<Planet>(planets, LocalizationDepth.Shallow);
_localizer.Localize<Planet>(planets, LocalizationDepth.OneLevel);
_localizer.Localize<Planet>(planets, LocalizationDepth.Deep);
```

Finally, if possible, LocalizationDepth.Deep should detect and avoid self reference loops and circular dependencies similar to JSON.net's `PreserveReferencesHandling = PreserveReferencesHandling.Objects`.

**Comment by ghost** [comment]
Want to do that.

**Comment by ghost** [comment]
Validate my interpretation of this please. Deserlizing an instance of `TestClass`

### Shallow
```
    public class TestClass : ILocalizable
    {
        [Localized]
        public string Name { get; set; } //Will be localized

        public TestClass2 Another { get; set; } //Will not be localized (notice that this is localizable)

        public IEnumerable<TestClass2> AnotherCollection { get; set; } //Will not be localized (notice that this is localizable)

        public TestClass Self { get; set; } //Will not be localized (notice that this is localizable)

        public TestClass(string name)
        {
            Name = name;
        }
    }

    public class TestClass2 : ILocalizable
    {
        [Localized]
        public string Name { get; set; } //Will not be localized

        public TestClass Test { get; set; } //Will not be localized
    }
```

### One Level
```
    public class TestClass : ILocalizable
    {
        [Localized]
        public string Name { get; set; } //Will be localized

        public TestClass2 Another { get; set; } //Will be localized, including props will be shallow

        public IEnumerable<TestClass2> AnotherCollection { get; set; } //Will be localized, including props will be shallow

        public TestClass Self { get; set; } //Will be localized (if it doesn't refer to self), including props will be shallow

        public TestClass(string name)
        {
            Name = name;
        }
    }

    public class TestClass2 : ILocalizable
    {
        [Localized]
        public string Name { get; set; } //Will be localized

        public TestClass Test { get; set; } //Will not be localized (notice that this is localizable)
    }
```

### Deep
```
    public class TestClass : ILocalizable
    {
        [Localized]
        public string Name { get; set; } //Will be localized

        public TestClass2 Another { get; set; } //Will be localized

        public IEnumerable<TestClass2> AnotherCollection { get; set; } //Will be localized

        public TestClass Self { get; set; } //Will be localized (if it doesn't refer to self), including props will be as well till shallow is reached!

        public TestClass(string name)
        {
            Name = name;
        }
    }

    public class TestClass2 : ILocalizable
    {
        [Localized]
        public string Name { get; set; } //Will be localized

        public TestClass Test { get; set; } //Will be localized if it doesn't refer to (test class), including proprs will be as well till a shallow level is reached!
    }
```

**Comment by ghost** [comment]
Here is actually part of the code without the deep part.

```
        private void LocalizeItem<T>(T item, string languageCode, LocalizationDepth depth = LocalizationDepth.Shallow)
            where T : class, ILocalizable
        {
            //TODO(t): Cache the three following steps/props (in memory ConcurentDictionary) for better pref (if possible)
            foreach (var property in typeof(T).GetTypeInfo().DeclaredProperties)
            {
                if (property.IsDefined(typeof(LocalizedAttribute)))
                {
                    LocalizeProperty(item, property, languageCode);
                }
                else if (depth != LocalizationDepth.Shallow)
                {
                    if (typeof(ILocalizable).IsAssignableFrom(property.DeclaringType))
                    {
                        LocalizeItem(property.GetValue(item, null) as ILocalizable, languageCode, LocalizationDepth.Shallow);
                    }
                    else if (typeof(IEnumerable<ILocalizable>).IsAssignableFrom(property.DeclaringType))
                    {
                        foreach (var child in property.GetValue(item, null) as IEnumerable<ILocalizable>)
                        {
                            LocalizeItem(child, languageCode, LocalizationDepth.Shallow);
                        }
                    }
                }
            }
        }

        private void LocalizeProperty<T>(T item, PropertyInfo propertyInfo, string languageCode)
            where T : class, ILocalizable
        {
            var propertyValue = propertyInfo.GetValue(item)?.ToString();
            if (string.IsNullOrWhiteSpace(propertyValue))
            {
                return;
            }

            if (!TryDeserialize(propertyValue, out var localizedContents))
            {
                return;
            }

            var contentForLanguage = GetContentForLanguage(localizedContents, languageCode);
            propertyInfo.SetValue(item, contentForLanguage, null);
        }
```

**Comment by mehalick** [comment]
I love it, looks great!

**Comment by ghost** [comment]
What do you think of this as an initial implementation of deep localization?

https://github.com/Tarek-Samy/Xaki/blob/deep-localization/src/Xaki/ObjectLocalizer.cs

I still have to do some light testing, but I think that should be it. Unless you have an objection to using `ReferenceEqual` to check for object equality/sameness. But what else would you use to check if two reference types are actually the same/equal.

My question would then become: In an object graph of reference types, what determines equality/sameness anyway?

Of course we could take a different approach and flatten the graph, take out duplicate items (again how 
 will you decide what is duplicate and what is not with reference types generically?), then localize the flat graph (now list), and add localized values in original places.

**Comment by mehalick** [comment]
If `ReferenceEquals()` works then I'd imagine that's a suitably efficient way to do it. You might not need to compare `GetType()` first although it can't hurt and probably wouldn't slow it down. Maybe test and perf test with and without and remove if it's not needed.

An alternative, but perhaps a hack, would be to check the first `LocalizedAttribute` property and if it's already localized then break/skip.

I also thought about adding metadata to objects, for example just to mark if it's already been localized, but I don't know if that's even possible. There is a concept of metadata programming in C# but I've never explored it.

One final thought, if deep localization is fast enough maybe we should make that the default instead of shallow? I've been thinking about that because the most common use case would actually be deep localization, so perhaps that can be the default. Only in situations where it's not the desired behavior and/or you're looking for maximum perf then you could use an overload that accepts `Shallow` or `OneLevel`. The reason why it wasn't the default in `OrangeJetpack.Localization` was there's no protection against circular references so the default would constantly shoot you in the foot.

**Comment by ghost** [comment]
Yeah, I thought a lot about `GetType` and I know it's not really needed, the code did read better to me so I kept it. When I do the PR, feel free to remove it.

I refuse to see this hack :), though it's an interesting approach I have to say that.

For the metadata talk, unless I'm mistaken you are practically talking about attributes in that case because meta programming (to the best of my knowledge) is

- Attributes
- CodeGeneration (like CodeDom)
- Reflection
- Dynamic Language Runtime (ExpandoObject/Dynamic)

You used two already! So anyways, It's doable via Attributes and Reflection in our case. Simplest way to do it (not best) would basically be
- Introduce a new attribute `Localizable` with a `IsLocalized` property flag
- Attach this attribute to all classes that implement `ILocalizable` interface at runtime. People will tell you that you can't do that, but you can! Anything is possible with reflection
- On item localization if localization is successful mark `IsLocalized` true
- On item localization read the attribute and If `Islocalized` is true skip

Instead of second step you can just use `TypeDescriptor.AddAttributes` on step 3.

But again if you really think about what you are doing, a simple list of the localized items make more sense than all that reflection.

If you actually mean something else by metaprogramming in C#, let me know!

**Comment by ghost** [comment]
Or you know instead of `Ilocalizable` attribute that we have now, we could switch to an attribute based system. So we would add a `Localizable` attribute on classes that are localizable. And of course that attribute would have a flag that is set on item localization.

**Comment by mehalick** [comment]
Implemented in v0.0.6
